### ISSUE TITLE
Large number of duplicates generated by streamloader

### ISSUE THREAD

**Issue by talyz** [snowplow-incubator/snowplow-bigquery-loader#168]
When setting up our Snowplow pipeline we've noticed a large amount of duplicate events being generated. For example, pushing 10000 events with a unique user id field gave me 6420 duplicate events and 3580 unique ones. Sometimes we even get a few hundred triple events, sometimes all events are unique.

After investigating the issue, the most likely source of these seems to be the streamloader:
1. Only the streamloader's CPU usage is abnormally high when these duplicates are generated
2. Turning off the mutator and repeater does not fix the issue
3. The enrichment reports the correct amount of events in its metrics output

It's hard to tell exactly why this is happening, since the streamloader doesn't provide any logging or debug output.

We're using version `0.6.2`.

**Comment by dilyand** [comment]
Hey @talyz, yes these duplicates are most likely coming from streamloader, when it’s run in a multi-node setup. This component is still with experimental status, as we note in the documentation. We can’t yet commit to a firm release date but we’re planning to introduce the first “prod-ready” version in 0.7.0 and duplicates are one of the things we want to address. I’m hesitant to link to the 0.7.0 milestone as it’s still work in progress but it will be the best place to follow.

The dataflow-based loader doesn’t produce duplicates but I believe from what you’ve said previously that it’s not an option for your setup.

As a work around, are you able to run the streamloader on a single worker without hitting a performance ceiling? That should eliminate the dupes.

**Comment by talyz** [comment]
Hi @dilyand! Yes, I know that it's still experimental, but it seemed to fit us best, which is why we're trying it, reporting issues and helping out where we can :)

The strange thing is that we're only running a single instance of the loader, but we're still getting duplicates. The loader is using quite a few threads in parallel, though. Currently, the whole pipeline is run by a single VM, one systemd service for each component, since we're still experimenting.

**Comment by dilyand** [comment]
That's much appreciated!

We've tested streamloader on a single node as a Kubernetes Engine deployment on GCP, and in those tests we did not see any duplicates in tests with 10k events. We only (expectedly) saw them when we allowed the pods to autoscale. It's interesting to hear that in other environments the behaviour might be different.

**Comment by talyz** [comment]
After looking deeper into the metrics on pub/sub, it seems this is caused by the messages being resent due to the ack deadline being exceeded. It happens to both the enrichment and the loader subscription (and likely the others too, but I haven't tested that). I've raised all the ack deadlines to 10 minutes, but that doesn't seem to be respected, since I'm still seeing loads of expired messages even though processing a whole batch only takes a couple of minutes.

Looking at the code, it looks like messages aren't acknowledged when they're received, but when they're done being processed. Acknowledging them as early as possible would probably help quite a lot. However, since pub/sub doesn't guarantee that a message will only be delivered once, the pipeline components would also have to keep track of what they've already processed for this to be entirely solved.

**Comment by dilyand** [comment]
Superseded by https://github.com/snowplow-incubator/snowplow-bigquery-loader/issues/192
